File: aeongseo/BOJ/CPP/boj10828.cpp
/*** 10828. 스택 (D2) ***/

#include<iostream>
#include<string>
using namespace std;

int main() {
	int N;
	cin >> N;

	int top = -1;
	int stack[10000];

	for (int i = 0; i < N; i++) {
		string order;
		int x;
		cin >> order;
		if (order == "push") {	// push일 때 정수 x를 더 받고 top 증가 후 stack에 저장
			cin >> x;
			top++;
			stack[top] = x;
		}
		else if (order == "pop") {	// pop일 때 top=-1이면(스택이 비었으면) -1을 출력, -1이 아니면 top 1 줄이고 맨위 stack 값 출력
			if (top == -1) {
				cout << -1 << '\n';
			}
			else{
				top--;
				cout << stack[top + 1] << '\n';
			}
		}
		else if (order == "size") {	// size일 때 top+1 출력
			cout << top + 1 << '\n';
		}
		else if (order == "empty") {	// empty일 때 top=-1이면(스택이 비었으면) 1 출력, -1아니면 0 출력
			if (top == -1) {
				cout << 1 << '\n';
			}
			else {
				cout << 0 << '\n';
			}
		}
		else if (order == "top") {	// top일 때 top=-1이면(스택이 비었으면) -1 출력, -1 아니면 맨위 stack 값 출력
			if (top == -1) {
				cout << -1 << '\n';
			}
			else {
				cout << stack[top] << '\n';
			}
		}
	}

	return 0;
}

File: seonil/BOJ/PYTHON/boj10828.py
"""
10828. 스택
"""

import sys

N = int(input())

# 스택을 구현하기 위한 고정 크기 배열 (최대 10000개까지 저장 가능)
stack = [0] * 10000
top = -1  # 스택의 가장 위(top) 원소의 인덱스. 비어있으면 -1

for i in range(N):
    command = sys.stdin.readline().split()  # 명령 입력 (push X / pop / size / empty / top)

    if command[0] == 'pop':  # pop 연산
        if top == -1:  # 스택이 비어있으면
            print(-1)
        else:  # 스택이 비어있지 않으면
            pop = stack[top]  # 스택에서 top 위치의 값 꺼내기
            top -= 1  # top을 한 칸 내림
            print(pop)  # 꺼낸 값 출력

    elif command[0] == 'size':  # size 연산
        print(top + 1)  # size = (top 인덱스 + 1)

    elif command[0] == 'empty':  # empty 연산
        if top == -1:  # 스택이 비어있으면
            print(1)
        else:  # 비어있지 않으면
            print(0)

    elif command[0] == 'top':  # top 연산
        if top == -1:  # 스택이 비어있으면
            print(-1)
        else:  # 스택이 비어있지 않으면
            print(stack[top])  # top 위치의 값 출력

    else:  # push X 연산
        x = int(command[1])  # 푸시할 값을 정수 형태로 저장
        top += 1  # top 인덱스를 한 칸 올림
        stack[top] = x  # 스택의 해당 위치에 값 저장(push)